### PR TITLE
Breaking: Plugins envs and config removal (fixes #4782, fixes #4952)

### LINKED ISSUE #4782: Custom environments in plugins
We discussed having custom environments a while back, but we got stuck due to the relationship between rules and environments. (Environments used to be able to enable/disable rules.) Since environments now just apply globals and parser options, I'd like to revisit allowing custom environments from plugins. This would relieve some of the pressure we get to provide an environment for every possible library.

## Defining

The idea is to export an `environments` property from plugins, just like we export `rules`:

```js
// your plugin
module.exports = {
    rules: {},
    environments: {
        custom: {
            globals: { "$": true }
            parserOptions: { ecmaVersion: 6 }
        }
    }
};
```

## Using

You'd use the same convention to load environments from plugins as you do for loading rules from plugins:

In config files:

```yaml
plugins:
  - myplugin
env:
  es6: true
  myplugin/custom: true
```

In comments:

```js
/* eslint-env myplugin/custom, es6*/
```

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29317762-custom-environments-in-plugins?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### LINKED ISSUE #4952: Remove default rule configuration for plugins
In updating plugins to support environments, I found that we're still supporting `rulesConfig` in plugins. I thought we had removed this feature in 1.0.0, but apparently not.

In carrying through the spirit of not forcing anything on anyone, I think it would be best to remove `rulesConfig` support and replace it with named configurations in plugins (#3659). That way, a plugin can still export rule configuration, but instead of forcing it on users, users can opt-in (just as they do with `eslint:recommended`).

Thoughts?

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/29882755-remove-default-rule-configuration-for-plugins?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F282608&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by nzakas**
This implements environments for plugins and also removes `rulesConfig` from plugins (because it was easier to do that than to fix all the failing tests for `rulesConfig` with this change).

### CODE REVIEW

File: lib/config.js [hunk 1]
```diff
@@ -15,7 +15,7 @@
 var path = require("path"),
     ConfigOps = require("./config/config-ops"),
     ConfigFile = require("./config/config-file"),
-    util = require("./util"),
+    Plugins = require("./config/plugins"),
```
Comment (ilyavolodin) [inline lib/config.js]: Why capital "P"? Doesn't look like this is using `new` anywhere

File: lib/config/config-ops.js [hunk 2]
```diff
@@ -12,7 +12,7 @@
 //------------------------------------------------------------------------------
 
 var debug = require("debug"),
-    environments = require("../../conf/environments"),
+    Environments = require("./environments"),
```
Comment (ilyavolodin) [inline lib/config/config-ops.js]: Same as above. I though we only capitalize objects that need to be newed. We even have a rule for this.

File: lib/config.js (hunk 1 above)
Comment (nzakas) [inline lib/config.js]: This is the way I've been doing these in `./config`. See `ConfigFile`, `ConfigOps`, etc.

I was having a hard time distinguishing between object instances and decided to try and clarify by making them look more like classes with static methods.

**Comment by nzakas** [comment]
Had to push to test. Once again, my local environment is borked for some reason (freezes on "Validating Rules"). I can't figure out what's going wrong, so I'll be pushing updates to validate fixes.

**Comment by gyandeeps** [comment]
@nzakas thats happening on AppVeyor also. I had to cacel the run as it was running for 1 hr.
https://ci.appveyor.com/project/nzakas/eslint/build/139

**Comment by nzakas** [comment]
Weird. Must be Windows-specific issue?

**Comment by nzakas** [comment]
I'm able to get past that point on master, though, so something is different. :-/

**Comment by ilyavolodin** [comment]
Tried this locally, and getting the same issue, so it's not just you.

**Comment by gyandeeps** [comment]
On my machine also the execution freezes for unit test. But after a quick discetion of the problem, there are unit test failing for `cli-engine`. Most of them are where its not able to find a particular plugin or config module. 
Guess: My guess is that the error is getting silently eaten somewhere. Here is where the error comes up : https://github.com/eslint/eslint/blob/issue4782/lib/config/plugins.js#L84

**Comment by nzakas** [comment]
Ah interesting. I'm still working on the errors, I'll see if that resolves the problem.

**Comment by nzakas** [comment]
I think I've fixed the errors, so pushing that now. I still can't get `npm test` to run locally. :( Be on the lookout for anything on this diff that might be the source of the problem. I can run `npm test` on the current master without any problems, so it must be something I changed in this branch.

**Comment by nzakas** [comment]
Still failing tests. :( Investigating.

**Comment by nzakas** [comment]
With @gyandeeps's help, code coverage is where it needs to be. :) If anyone wants to take one last look at this before I merge, please do so.

**Comment by ilyavolodin** [comment]
LGTM.

**Pull request merged**